_posts/2016-08-26-babili.md

```diff
@@ -0,0 +1,183 @@
+
+TL;DR: Babili can output to ES2015+ code so you don't need to transpile and then minify if it's not necessary. Babili is also modular/flexible (it's a babel preset and thus supports user plugins) and can be used as a preset or cli tool. Babili should also be able to do ES2015+ specific optimizations.
+
+## Prounciation
+
+```bash
+# sounds like "bah billy"
+say babili
+```
+
+
+> If you can't remember the name, babel-minify works too.
+
+## Why minify?
+
+[Minification](https://en.wikipedia.org/wiki/Minification_(programming)) removes unncessary characeters without changing its functionality. The basics of minification remove things like comments, whitespace, newlines, and extra parentheses. You can also simplify code, transform code to smaller equivalents, remove unused code, etc.
+
+Minification is useful because it reduces the amount of code transferred from the server to the client. Because users download less, the page can load faster.
+
+## Current minifiers
+
+Tools such as [uglify](https://github.com/mishoo/UglifyJS2) don't currently support targeting the latest version of Javascript ([yet](https://github.com/mishoo/UglifyJS2/issues/448)).
+
+We currently to use tools like Babel to compile ES2015 code down to ES5 code to support older browsers. Then we use something like uglify to cut down on the bundle size.
+
+As browsers implement more ES2015 features and we drop support for older browser versions, there is a sliding window of the version of javascript you write in and the target javascript version you compile/minify to. However because uglify cannot parse ES2015, you would still have to compile down to ES5 anyway.
+
+## Babili
+
+That's where Babili comes in.
+
+Babili is ES2015+ aware because it is built using the Babel toolchain. More specifically, it is a set of babel plugins + a preset just like with the `es2015` preset.
+
+Like described in [Not Born to Die](http://babeljs.io/blog/2015/02/15/not-born-to-die) when Babel was renamed from 6to5, Babel was originally a specific transpiler for ES2016 to ES5. A transpiler is just a compiler so the same concepts to transform arrow functions can work to minify code.
+
+## Example
+
+When it's possible to only target browsers that support newer ES features, code sizes can be smaller because you don't have to transpile and then minify.
+
+For example: say we want to minify a file that contains an ES2015 class and we want to target the latest version of Chrome.
+
+```js
+class Mangler {
+  constructor(program) {
+    this.program = program;
+  }
+}
+// need this since otherwise Mangler isn't used
+new Mangler(); 
+```
+
+Before, we might run Babel to transpile [the class into a function](http://babeljs.io/docs/plugins/transform-es2015-classes) and run uglify on the compiled code to send to the browser.
+
+```js
+// ES2015 code -> Babel -> Uglify/Babili -> Minified ES5 Code
+var Mangler=function a(b){_classCallCheck(this,a),this.program=b};Mangler();
```

Comment:
It's even less fair if you don't actually instantiate (`new Mangler()`) but have a superclass or any class member other than the constructor. Uglify will still leave the constructor because the transform converts the class into an IIFE, and uglify can't remove live code, even if it's a pure function with unused result.